File: solutions/fsharp/proverb/1/Proverb.fs
module Proverb

let recite (input: string list): string list =

    match input with
    | [] -> input
    | head::_ ->
        [
            for pair in Seq.pairwise input do
                let (fst: string, snd) = (fst pair, snd pair)
                yield sprintf "For want of a %s the %s was lost." fst snd
            yield sprintf "And all for the want of a %s." head
        ]

File: solutions/csharp/proverb/1/Proverb.cs
public static class Proverb
{
    public static string[] Recite(string[] subjects)
    {
        IEnumerable<(string, string)> Pairs()
        {
            for (var idx = 0; idx < subjects.Length - 1; idx++)
                yield return (subjects[idx], subjects[idx + 1]);
        }

        IEnumerable<string> Loop()
        {
            foreach (var pair in Pairs())
                yield return $"For want of a {pair.Item1} the {pair.Item2} was lost.";
            yield return $"And all for the want of a {subjects[0]}.";
        }

        return subjects.Length == 0
            ? System.Array.Empty<string>()
            : Loop().ToArray();
    }
}